RFC: User Defined Type Functions (#45)
luau-lang/rfcs: docs/user-defined-type-functions.md

Reviewer: Something not talked about here is the fact that you're able to make a class type in userspace now, which I think is weird?

Reviewer 2: We don't need to make this possible at all, and indeed probably should not.

Author: on the most recent commit, i added more specifications on how `typelib`s can be created and it doesn't include a method of creating new class types

- Enable more precise type definitions that can capture complex relationships and constraints
- Facilitate the creation of reusable and composable libraries
- Enhance code safety and maintainability with better type-level programming

The expected outcome is a more versatile type system that can adapt to a wider range of programming patterns, ultimately improving the Luau developer experience.

## Design

Type functions wil be defined with the following syntax:
```luau
type function f(...)
    -- implementation of the type function
end
```

For instance, the `rawget` type function can be written as:
```luau
type function rawget(tbl, prop)
    return tbl.getProps()[prop.getValue()]

    -- returning nothing / nil means that this type function will fail to reduce
end

type Person = {
    name: string,
    age: number
}

type ty = rawget<Person, "name"> -- ty = string
```

> To allow Luau developers to modify the runtime values of types in type functions, this RFC proposes introducing a new userdata called `lType`.  An `lType` object is a runtime representation of all types within the program and provides a basic set of library methods that can be used to modify types. As such, under the hood, each argument of a type function is serialized into a userdata called `lType`. Moreover, they are *only accessible within type functions* and are *not a runtime type for other use cases than type functions*. 

Reviewer 2: I don't think we should actually be calling the userdata `lType`. Certainly, the _header_ has that name because that's consistent with how we name Luau VM headers (a convention adopted from Lua itself), but the user-facing name should just be `type` or `types`. `type` would be most consistent with `string`, `table,` buffer`, but it's a bit thorny because it would mean shadowing the global function `type` that is already present in the VM. Possibly a better option is `luautype`, or maybe someone will have alternative suggestions for the library name.

Author: I should've mentioned that lType was a placeholder for the name in the RFC. I was also concerned with just calling it `type`. I think `luautype` is a good name.

Reviewer 3: Having luautype is strange for this being inside the Luau language.

Reviewer 4: What about a variation of `metatype`?

Author: we've decided to go with `typelib`!

> To allow Luau developers to modify the runtime values of types in type functions, this RFC proposes introducing a new userdata called `lType`.  An `lType` object is a runtime representation of all types within the program and provides a basic set of library methods that can be used to modify types. As such, under the hood, each argument of a type function is serialized into a userdata called `lType`. Moreover, they are *only accessible within type functions* and are *not a runtime type for other use cases than type functions*. 

Reviewer: Should be called `type`.

Author: > I don't think we should actually be calling the userdata lType. Certainly, the header has that name because that's consistent with how we name Luau VM headers (a convention adopted from Lua itself), but the user-facing name should just be type or types. type would be most consistent with string, table, buffer, but it's a bit thorny because it would mean shadowing the global function typethat is already present in the VM. Possibly a better option isluautype`, or maybe someone will have alternative suggestions for the library name.

what do you think about calling it `luautype`?

Reviewer: `type` is better. I don't think we should be concerned about shadowing the builtin global, the solution there is not to have a global named `type`.

Reviewer 2: Indeed, I do prefer `type` to `luautype` certainly, and frankly, I'm not sure anyone would even want to use `type` inside of a type function _anyway_ (it won't be interesting, it'd always be `userdata` under this proposal, or always `table` under the alternative proposal).

Author: what about later down the line when we want to add type checking to type functions? Theoretically, we would still want users to define custom types in type functions, right?

<details><summary>lType library methods (dropdown)</summary>
Note: methods under a different type heading (ex: `Singleton`) imply that the methods are only available for those types. At the implementation level, there is a check to make sure that the type-specific methods are being called on the correct types (e.g, for `getIndexer()`, assert that `isTable()` is true).

#### Any

> | Function Declaration | Return Type | Description |

Reviewer: I don't see anything about union types or intersection types.

Reviewer: Or generics, or type function instances.

Reviewer: Negation types.

Reviewer: Error suppressing types?

Reviewer 4: Type packs as well!

Reviewer 2: I think we'll probably consider type pack functions to be out of scope for the initial version, but if people have strong use cases for them, we can certainly consider it now.

Author: > I don't see anything about union types or intersection types.
> Negation types.
> Error suppressing types?

Added!

> Or generics, or type function instances.

I didn't really see an use case for this, so I didn't bring it up during my revision. do you think it's a must have for at least the initial version?

Author: > Type packs as well! typeFunc<(A, B), (C, D)>
And variadic types. typeFunc<...A>

Following Aaron's comment, the initial version only includes the ability to add type packs/variadic types to an instance of a function type, but doesn't have anything about the type functions themselves taking in type packs/variadic types

| ------------- | ------------- | ------------- |

> | `isString()` | `boolean` | returns true if self is of type `string` |

Reviewer: What's the reason for using `isT()` as opposed to using tagged unions?

```luau
type type = 
  | { type: "primitive", primitive: "number" }
  | { type: "primitive", primitive: "string" }
  | { type: "table", props: { [string]: type }, indexer: { key: type, value: type }? }
  | ...
```

Reviewer: I would even want to have something like
```luau
export type numbertype = { type: "primitive", primitive: "number" }
export type stringtype = { type: "primitive", primitive: "string" }
export type tabletype = { type: "table", props: { [string]: type }, indexer: { key: type, value: type }? }
...
```

so that e.g. you are able to constrain the domain of your type functions.

| `isnumber()` | `boolean` | returns true if self is of type `number` |

> | `isboolean()` | `boolean` | returns true if self is of type `boolean` |

Reviewer: `boolean` is equivalent to `true | false`.

> | `isboolean()` | `boolean` | returns true if self is of type `boolean` (e.g. true or false) |

Reviewer: Sorry, I didn't mean to make a note that `boolean` is equivalent to `true | false`. What I meant to say was there's both `isboolean()` and `isbooleansingleton()` below, so the former felt redundant.

I guess what I'm really asking here is, when a type argument is marshalled into a type function, what are we _really_ passing in? e.g.

```lua
type T = { p: string } & { q: number }
type U = { p: string, q: number }
```
Is `my_type_fun<T>` passed the same structure as `my_type_fun<U>`?

Reviewer 2: I would imagine they would get different types because type functions would work syntactically, but an alternative approach would be using normalization to canonize the type. Unfortunately, normalization doesn't actually normalize tables very much as-is though, so that would require a bunch of additional work on our part to expand the scope of normalization. One possibility here is that we take it syntactically in whatever form it is, and we could consider adding normalization (or some of the normalization APIs, like `isIntersectionInhabited`) to the API surface for the types library, but that is a bit scary.

Reviewer: I kind of want to avoid making promises on this front at least until egraphs is fully implemented, then we can always pass in a type that's guaranteed to be most simplest, for some definition of simplest.

Then users won't have to worry about dealing with cases like `string & ~"a"` vs `{ p: number } & { q: string }`.

| `isTable()` | `boolean` | returns true if self is of type `table` |
| `isThread()` | `boolean` | returns true if self is of type `thread` |
| `isFunction()` | `boolean` | returns true if self is of type `function` |
| `isBuffer()` | `boolean` | returns true if self is of type `buffer` |
| `isNil()` | `boolean` | returns true if self is of type `nil` |
| `isClass()` | `boolean` | returns true if self is of type `class` (do we need this?) |
| `isBooleanSingleton()` | `boolean` | returns true if self is a boolean singleton |
| `isStringSingleton()` | `boolean` | returns true if self is a string singleton |

> | `isA(arg: lType)` | `boolean` | returns true if arg is the same type as self |

Reviewer: What motivated this one?

Author: the motivation behind `isa()` was to provide a way for developers to compare the relationships of different types. I saw [your comment](https://github.com/luau-lang/rfcs/pull/45#discussion_r1651707952) and thought that it was a better idea than what I proposed, so I am planning on making the change to your idea on my revision.

> | `isa(arg: lType)` | `boolean` | returns true if arg is the same type as self |

Reviewer: How about `__eq` and using `==` to decide equality.

You can also even use `>` and `<` to decide subtyping, so `number < (number | string)` returns true.

Reviewer 2: If we're doing the latter, it should be `<=` and `>=` since subtyping is reflexive.

Reviewer: Good point. This is easy to do with just defining `__le`.

Author: this sounds like a great idea to me. I will add it to my next revision

#### Primitive

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `getType()` | `string` | returns either "nil", "boolean", "string", "thread", "function", "table", or "buffer" |

#### Singleton

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `getValue()` | `string` | returns either "true", "false", or a string singleton |

#### Table

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `getProps()` | `table` | returns a type representation of tables (e.g. {name = "John"} will return {[string] = "string"}) |

> | `getIndexer()` | `table` | returns a type representation of arrays (e.g. {1, "hi", 3} will return {[number] = "number" \| "string"}) |

Reviewer 2: Neither of these seem to be what I would expect them to be. The properties of a table are a bunch of key-value pairs where the key is a type and the value is a type. Likewise, the indexer and the result of indexing are both types.

Author: a recent commit includes this change!

> #### Class

Reviewer 2: userdata have properties and indexers too, but this proposal doesn't currently give us a way to access them.

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |

> | `getName()` | `table` | returns the name of self's class |

Reviewer 2: If this is returning a name, why is it `table`? Shouldn't it just be `string`?

Reviewer: This will be problematic with named types that I want to do IMO.

| `getParent()` | `lType` | returns lType userdata of self's parent |

</details>

> ### Implementation

Reviewer: Can probably be left out of the RFC, in my opinion.

Reviewer 2: There's a better implementation section that could be written that actually describes the interesting work here (implementing a `user` type function into the builtins that calls serialization and deserialization code and interacts with the VM itself to run user code in an established "type function" environment with the right libraries available and all of the type aliases and functions from the code available).

Author: In a recent commit, I've revised the implementation section to leave out specific details and include more details on the interactions with the VM

## Drawback

Type functions are handled at the analysis time, while `lType` is an implementation in the runtime. As a result, the proposed design causes the analysis time to be dependent on Luau's runtime to reduce user-defined type functions. This is generally discouraged as it is best to isolate the compile time, analysis time, and runtime from each other for the purpose of maintaining a clean separation of concerns, which helps minimize side effects and dependencies across different phases of the program execution and improves the modularity of the compiler. Overlaps between the analysis time and runtime can lead to code that is more complex and harder to manage, potentially increasing the risk of bugs and making the outcomes less predictable.

The build / analysis times will also be negatively impacted by this feature as reducing type functions takes variable amount of time based on the program. The larger the type function, the longer it will take to reduce it in the constraint solver.

## Alternatives

### `table` Runtime Representation

> Currently, the runtime representation of types is a userdata called `lType`. Another representation is to use the already-existing type in Luau `table`; instead of serializing types into `lType`, we can serialize them into tables with predefined properties. For instance, the representation for a string singleton `"abc"` could be `{type = "stringSingleton", value = "abc"}`. So instead of writing:

Reviewer: I would much rather have this runtime representation!

Reviewer 2: I'm curious as to why. I originally wrote it that way in the earliest prototype for type functions, but I directed Jeremy to implement it as a userdata because we are, fundamentally, embedding types from the type system into these VM instances and that's... really what userdata is _for_. I do think that the corresponding C++, if we go the userdata route, should probably be a little OO hierarchy for the types, which isn't quite what's described here. But I envisioned something akin to the instance hierarchy we embed from Roblox.

One question here is which of these would be easier to add typing for (i.e. actually type checking the type functions), and I think the userdata approach might be somewhat easier? But at the end of the day, they're probably both going to lean pretty heavily on refinements.

Reviewer: mmm. I guess it can go either way, though since I suggested using `==`/`>=`/`<=` for subtyping, the only way that'd realistically work is by using userdata. Let's go with that then.

```luau
type function isSingleton(t)
    if t:isStringSingleton() then
        return t
    end
end
```
developers could write:
```luau
type function isSingleton(t)
    if t.type == "stringSingleton" then
        return t
    end
end

> ```

Reviewer 2: Both of these type functions are missing a final return value for the failing cases.

Author: functions have implied return value of nil, so my intention was to show that these type functions will fail to reduce if it returns nil / doesn't specify a return value. This was also mentioned in the "Design" section when I wrote an example for `rawget`.

Reviewer: I was hoping we would use `error("reason")` instead of `return nil` or `return`. The error function would then get displayed in type errors.

Reviewer 2: Functions do not have an implied return of `nil`. Functions that do not have a return statement do not return anything, which is distinct from returning a `nil` value. You can see this by looking at the semantics of `select("#", ...)` on a function that has `return nil` vs no return.

> One of the major advantages of using a table-based representation for types in Luau is that it eliminates the dependency on the virtual machine for type analysis. However, this design requires developers to have a deeper understanding of the runtime representation. For example, under the proposed design, a new table type can be created using `lType.new("table")`, while under the alternative design, tables must be declared with attributes: `{type = "table", props = {}, indexer = {}}`. This adds complexity to the developer experience and increases the chance of making syntax errors in their program.

Reviewer 5: What prevents defining `lType.new` that returns a table?

Reviewer 2: Nothing at all. If we did choose a table representation for types instead of a userdata, we would still provide a whole library to be linked into the VM created in analysis to support writing type functions effectively.

Author: As per this [comment](https://github.com/luau-lang/rfcs/pull/45#issuecomment-2187526252), `lType` userdata is only registered when reducing UDTFs, so any other uses of `lType` outside of type functions will be a call to invalid / nonexistant class.

Reviewer 3: I'm just saying that this is kind of presented as a reason to go with userdata instead of tables.
There might be other reasons, but tables feel like a more natural primitive.
Userdata is a more rigid structure and reflection is likely to increase complexity (I don't think we can connect to TypeVariant directly here).

> One of the major advantages of using a table-based representation for types in Luau is that it eliminates the dependency on the virtual machine for type analysis. However, this design requires developers to have a deeper understanding of the runtime representation. For example, under the proposed design, a new table type can be created using `lType.new("table")`, while under the alternative design, tables must be declared with attributes: `{type = "table", props = {}, indexer = {}}`. This adds complexity to the developer experience and increases the chance of making syntax errors in their program.

Reviewer 2: > One of the major advantages of using a table-based representation for types in Luau is that it eliminates the dependency on the virtual machine for type analysis.

This is not true. In order to run type functions, analysis will clearly always need to depend on the virtual machine, regardless of how we choose to represent types in the runtime for those type functions.

> One of the major advantages of using a table-based representation for types in Luau is that it eliminates the dependency on the virtual machine for type analysis. However, this design requires developers to have a deeper understanding of the runtime representation. For example, under the proposed design, a new table type can be created using `lType.new("table")`, while under the alternative design, tables must be declared with attributes: `{type = "table", props = {}, indexer = {}}`. This adds complexity to the developer experience and increases the chance of making syntax errors in their program.

Reviewer 2: Nothing in the earlier portion of this RFC actually describes `type.new` or any other global functions that would be available. I'm not sure what semantics you're imagining for `type.new` based on this comment, but I imagined we would have functions to construct the various sorts of types that exist, so e.g. `type.makeunion`, `type.maketable`, `table.makeintersection`, `table.makeuserdata`, etc. (feel free to bikeshed names, could be `newunion`, or `freshtable`, or maybe some other stuff, the point is just to convey the idea here).

Author: a recent commit includes more specifications on how instances of typelibs can be created and what types are supported (currently adding support for generic types)

To allow Luau developers to modify the runtime values of types in type functions, this RFC proposes introducing a new userdata called `lType`.  An `lType` object is a runtime representation of all types within the program and provides a basic set of library methods that can be used to modify types. As such, under the hood, each argument of a type function is serialized into a userdata called `lType`. Moreover, they are *only accessible within type functions* and are *not a runtime type for other use cases than type functions*. 

<details><summary>lType library methods (dropdown)</summary>

> Note: methods under a different type heading (ex: `Singleton`) imply that the methods are only available for those types. At the implementation level, there is a check to make sure that the type-specific methods are being called on the correct types (e.g, for `getIndexer()`, assert that `isTable()` is true).

Reviewer: Don't forget to rename these.

## Motivation

The primary motivation for introducing user-defined type functions is to extend the flexibility of the type system and empower developers to create more expressive and type-safe libraries. Current limitations in the type system restrict custom type manipulations, which are crucial for creating advanced patterns and abstractions. By supporting user-defined type functions, we aim to:
- Enable more precise type definitions that can capture complex relationships and constraints
- Facilitate the creation of reusable and composable libraries
- Enhance code safety and maintainability with better type-level programming

The expected outcome is a more versatile type system that can adapt to a wider range of programming patterns, ultimately improving the Luau developer experience.

## Design

Type functions wil be defined with the following syntax:
```luau
type function f(...)
    -- implementation of the type function
end
```

For instance, the `rawget` type function can be written as:
```luau
type function rawget(tbl, prop)
    print("First parameter should be a table") -- produces a warning

>     if (~tbl.istable()) then

Reviewer 6: What is this `~` syntax? It's not specified in the RFC

>     if (~tbl.istable()) then

Reviewer 6: Surely `tbl:istable()`? Even if these don't have actual metatables it seems inconsistent with general Luau expectations not to

type function neverending(t)
    return neverending(t) -- note: we use parenthesis here because we are passing in the runtime value of types, not static values
end
```
We have considered adding an user-configured execution limit based on time or instruction count for reducing type functions where if a type function does not finish executing under the constraint, it fails to reduce. However, time-based timeouts are dependent on CPU; programs that type check on fast CPUs may not type check on slower CPUs. Similarly, instruction-based timeouts are dependent on the compiler version; programs that type check on versions of compilers where they produce less instructions may not type check on compilers that do not carry the same optimizations. We will not (and probably never) allow type functions to call regular functions for the sake of sandboxing the runtime and analysis.

To give warnings, developers can use `print()` with custom warning messages. To fail reductions, developers can use `error()` with custom error messages. If nothing is returned by the type function, it will fail to reduce with the default message: "Failed to reduce \<Name\> type function with no return values".

> To allow Luau developers to modify the runtime values of types in type functions, this RFC proposes introducing a new userdata called `typelib`.  An `typelib` object is a runtime representation of all types within the program and provides a basic set of library methods that can be used to modify types. As such, under the hood, each argument of a type function is serialized into a userdata called `typelib`. Most importantly, they are *only accessible within type functions* and are *not a runtime type for other use cases than type functions*. 

Reviewer 6: ```suggestion
To allow Luau developers to modify the runtime values of types in type functions, this RFC proposes introducing a new userdata called `typelib`.  A `typelib` object is a runtime representation of all types within the program and provides a basic set of library methods that can be used to modify types. As such, under the hood, each argument of a type function is serialized into a userdata called `typelib`. Most importantly, they are *only accessible within type functions* and are *not a runtime type for other use cases than type functions*. 
```

Author: this was fixed in the most recent commit!

| `isboolean()` | `boolean` | returns true if self is of type `boolean` (e.g. true or false) |
| `istable()` | `boolean` | returns true if self is of type `table` |
| `isthread()` | `boolean` | returns true if self is of type `thread` |
| `isfunction()` | `boolean` | returns true if self is of type `function` |
| `isbuffer()` | `boolean` | returns true if self is of type `buffer` |
| `isnil()` | `boolean` | returns true if self is of type `nil` |
| `isclass()` | `boolean` | returns true if self is of type `class` (do we need this?) |
| `isbooleansingleton()` | `boolean` | returns true if self is a boolean singleton |
| `isstringsingleton()` | `boolean` | returns true if self is a string singleton |
| `isa(arg: typelib)` | `boolean` | returns true if arg is the same type as self |

#### Primitive

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `gettype()` | `string` | returns either "nil", "boolean", "string", "thread", "function", "table", or "buffer" |

#### Singleton

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |

> | `getvalue()` | `string` | returns either "true", "false", or a string singleton |

Reviewer 6: What happens if it's not a singleton? `nil`? Error?

### More Builtin Type Functions

Another alternative is to create more built-in type functions that cover a wider range of programming patterns. This approach clearly lacks the flexibility of fully user-defined type functions because type manipulation would still be limited to the predefined set of type functions designed by the Luau team. Furthermore, continuously expanding the set of built-in type functions leads to bloat and complexity within the language, making it harder to maintain.

## Future Works

> In the future, we could investigate adding type checking to user defined type functions. In Haskell's type families, this is done with _kinds_ (aka types of types). Similarly, we could look into introducing kinds to Luau, but for the purpose of this RFC.

Reviewer 2: ```suggestion
In the future, we could investigate adding type checking to user defined type functions. In Haskell's type families, this is done with _kinds_ (aka types of types). Similarly, we could look into introducing kinds to Luau, but, for the purpose of this RFC, we consider type checking user-defined type functions to be out of scope.
```

        if k == key then
            return v
        end
    end

    error("key not found!")
end

type Person = {
    name: string,
    age: number
}

type ty = rawget<Person, "name"> -- ty = string
```

Type functions operate on two stages: type analysis and runtime. When calling type functions at the type level (e.g. annotating a variable as a type function), angle brackets must be used, but when calling them at the runtime level (e.g. calling other type functions within type functions), parenthesis must be used. Declarations of type functions use parentheses because it defines the runtime operations on the runtime representation of types.

For the first iteration, the body of a type function will be sandboxed, and its scope will be limited, meaning it will be unable to refer statements defined in the outer scope, including other type functions. Additionally, type functions cannot be exported and will be limited on what globals/libraries they could call. The list of available globals/libraries in type functions is:

- global functions: `assert`, `error`, `next`, `print`, `rawequal`, `select`, `tonumber`, `tostring`, `type`, `typeof`, `ipairs`, `pairs`, `unpack`

> - math library

Reviewer 3: Can we specify that `math.randomseed` is reset before each type function execution entry?

Reviewer: How about not allowing `math.random` as well as other functions like it?

Reviewer 3: I think it's fine to include them as any pseudo-random generator can be implemented manually anyway.

You can even compute a hash of `tostring({})` to use as a non-deterministic seed value.

Reviewer 2: Added this explicitly!

Reviewer: This means type function instantiation can't be cached because you won't get the same resulting type per invocation.

| `copy(arg: type)` | `type` | returns a deep copy of the argument |

#### Negation

| Instance Methods | Type | Description |
| ------------- | ------------- | ------------- |
| `gettype()` | `type` | returns the `type` being negated |

#### Singleton

| Instance Methods | Return Type | Description |
| ------------- | ------------- | ------------- |
| `getvalue()` | `string \| boolean` | returns the string value of the  singleton |

#### Table

| Instance Methods | Return Type | Description |
| ------------- | ------------- | ------------- |
| `setprop(key: type, value: type?)` | `nil` | adds / overrides (if same key exists) a key, value pair to the table's properties; if value is nil, removes the key, value pair; if the key does not exist and the value is nil, nothing happens |
| `getprop(key: type)` | `type?` | returns the value associated with the key from the table's properties; if the key does not exists, returns nil |
| `getprops()` | `{[type]: type}` | returns a table of the table's properties |
| `setindexer(key: type, value: type)` | `nil` | sets the table's indexer with key type as the first argument and value type as the second argument |

> | `getindexer()` | `{key: type, value: type}?` | returns the table's indexer as a table; if the indexer does not exist, returns nil |

Reviewer: Just realized these aren't compatible with the eventual read/write properties that tables will have in the new solver. Indexers also needs read/write flags too.

Reviewer 2: This is true, but there's probably care that needs to be had here because we don't want people to be able to set up entirely disjoint read/write property types right now.

## Motivation

The primary motivation for introducing user-defined type functions is to extend the flexibility of the type system and empower developers to create more expressive and type-safe libraries. Current limitations in the type system restrict custom type manipulations, which are crucial for creating advanced patterns and abstractions. By supporting user-defined type functions, we aim to:
- Enable more precise type definitions that can capture complex relationships and constraints
- Facilitate the creation of reusable and composable libraries
- Enhance code safety and maintainability with better type-level programming

The expected outcome is a more versatile type system that can adapt to a wider range of programming patterns, ultimately improving the Luau developer experience.

## Design

Type functions will be defined with the following syntax:
```luau
type function f(...)
    -- implementation of the type function
end
```

For instance, the `rawget` type function can be written as:
```luau
type function rawget(tbl, prop)

>     if not typelib.type(tbl) == "table" then

Reviewer 4: nit: `not A == B` evaluates as `(not A) == B`, use `~=` instead

```suggestion
    if typelib.type(tbl) ~= "table" then
```

<details><summary>typelib library (dropdown)</summary>

Methods under a different type heading (ex: `Singleton`) imply that the methods are only available for those types. At the implementation level, there is a check to make sure that the type-specific methods are being called on the correct types. For instance, `getindexer()` asserts that `istable()` is true.

#### typelib
All attributes of newly created typelib are initialized with empty tables / arrays and `nil`. For instance, `typelib.newtable()` initializes its properties with an empty table and index / index result type as `nil`. Additionally, all arguments are passed by references.

| Instance Attributes | Type | Description |
| ------------- | ------------- | ------------- |
| `niltype` | `typelib` | an immutable runtime representation of the built-in type `nil` |
| `unknown` | `typelib` | an immutable runtime representation of the built-in type `unknown` |
| `never` | `typelib` | an immutable runtime representation of the built-in type `never` |
| `any` | `typelib` | an immutable runtime representation of the built-in type `any` |

> | `boolean` | `typelib` | returns an immutable runtime representation of the built-in type `boolean` |
> | `number` | `typelib` | returns an immutable runtime representation of the built-in type `number` |
> | `string` | `typelib` | returns an immutable runtime representation of the built-in type `string` |

Reviewer: ```suggestion
| `boolean` | `typelib` | an immutable runtime representation of the built-in type `boolean` |
| `number` | `typelib` | an immutable runtime representation of the built-in type `number` |
| `string` | `typelib` | an immutable runtime representation of the built-in type `string` |
```

| ------------- | ------------- | ------------- |
| `getnegation(arg: typelib)` | `typelib` | returns an immutable runtime representation of the negation of the argument; the argument cannot be `istable()`, `ismetatable` or `isfunction()` |
| `getstringsingleton(arg: string)` | `typelib` | returns an immutable runtime representation of a string singleton type of the argument |
| `getbooleansingleton(arg: boolean)` | `typelib` | returns an immutable runtime representation of a boolean singleton type of the argument |
| `getunion(arg: {typelib})` | `typelib` | returns an immutable runtime representation of union type of its argument |
| `getintersection(arg: {typelib})` | `typelib` | returns an immutable runtime representation of intersection type of its argument |
| `newtable(props: {[typelib]: typelib}?, indexer: {key: typelib, value: typelib}?)` | `typelib` | returns a mutable runtime representation of a `table` type |
| `newmetatable(props: {[typelib]: typelib}?, indexer: {key: typelib, value: typelib}?, metatable: typelib?)` | `typelib` | returns a mutable runtime representation of a metatable. `metatable` argument needs to be the same type as `typelib.newtable()` |

> | `newfunction(parameters: {typelib} \| typelib?, returns: {typelib} \| typelib?)` | `typelib` | returns a mutable runtime representation of a `function` type. Calling `newfunction(X)` will by default set `parameters` to `X` |

Reviewer: Alternative design space is

```luau
typelib.create("literal", "my string")
typelib.create("literal", true)
typelib.create("union", {my types})
typelib.create("table", props, indexer)
typelib.create("function", params, rets)
```

Reviewer 2: yeah, separate constructors vs one overloaded constructor. I personally prefer separate constructors.

Author: i would prefer separate constructors as well. for my project, i would like to keep it that way 🙂

> | `type(arg: typelib)` | `string` | returns the tag of the argument ("nil", "unknown", "never", "any", "boolean", "number", "string", "boolean singleton", "string singleton", "negation", "union", "intersection", "table", "metatable", "function", "class") |

Reviewer: Is this written as `typelib.type(ty)`? Because `typeof` covers this use case, you just need to call `lua_newuserdatatagged` and then add a `__type = "the type"` metamethod.

Reviewer: Though this leads to a conflict: `typeof(ty) == "string"` could be either the string primitive, or the `string` _type_. `ty:type() == "string"` might be better in this case.

Reviewer 2: I think the original thing I suggested informally was to actually just use `typeof`, but yeah, there's a bit of awkwardness there with `string` and that's why we wanted to do something else. I think `ty:is("string")` feels good though.

> | `type(arg: typelib)` | `string` | returns the tag of the argument ("nil", "unknown", "never", "any", "boolean", "number", "string", "boolean singleton", "string singleton", "negation", "union", "intersection", "table", "metatable", "function", "class") |

Reviewer: Separately, the `metatable` string shouldn't be in here.

> | `type(arg: typelib)` | `string` | returns the tag of the argument ("nil", "unknown", "never", "any", "boolean", "number", "string", "boolean singleton", "string singleton", "negation", "union", "intersection", "table", "metatable", "function", "class") |

Reviewer: Maybe we can even get rid of `"string singleton"` and `"boolean singleton"` and have both return just `"literal"`. Then you'd do this:

```luau
if ty:is("literal") and typeof(ty:getvalue()) == "string" then
  ...
end
```

Reviewer: This did expose a problem with the current API design, type refinements on a function call doesn't do anything. So `typeof(ty:getvalue() == "string"` followed by another call to `ty:getvalue()` still has type `"string" | "boolean"`.

Extending type refinements to be able to refine the type of a function call will be difficult, so you end up having to push the value onto a register, and then refine that. :/

Reviewer 2: I'd be okay with combining the singleton types into one tag, I think.

Author: I would be okay with that as well. In that case, I want to remove type.getstringsingleton / getbooleansingleton, and change it with: `type.getsingleton(arg: string | boolean)`

The key to making this type runtime work is the introduction of a `type` userdata that Luau's types can be serialized into. For clarity in this RFC, `type` (in code block) always refers to the userdata and type (the unadorned string) refers to the ordinary meaning of the word in the context of Luau (i.e.  the static type of a binding). An instance of the `type` userdata is a type runtime representation of a type within the program, and it provides a set of API calls that can be used to inspect and manipulate the type. As part of the _type runtime_, they are *only accessible within type functions* and are *not available during ordinary function runtimes*. Each type function can take an arbitrary number of `type` arguments, and can return exactly one `type` as a successful result during evaluation. Returning additional results will signal an error to the user for now, but a future RFC may revisit this to support user-defined type pack functions as well. We are currently excluding them since the scope of this RFC is already very large as-is, and even the built-in type function system has yet to find a compelling use for type pack functions.

> Because the name `type` clashes with the global function `type(...)`, the `type` userdata will set the `__call` metamethod will to the original `type(...)` function. This allows us to gracefully preserve the behavior of the built-in `type` function if it is necessary for developers, while still choosing the best possible name for the datatype itself. We considered alternative names like `luautype` `typelib` or `ltype`, but alternatives seemed to feel broadly worse when used in speech and writing.

Reviewer 3: My suggestion would be to use `types` as the library name.

I just don't think it's that great to alias multiple things on the same name and work-around with a `__call` override.
`type` function is part of the base library, so the type library will have to have a duplicate implementation of it.
Two implementations across such boundaries will be harder to support going forward.

If `types` in undesirable, I'm sure we can spend some effort to figure out a name :)

## Summary

The new type solver for Luau supports a limited number of built-in type functions that enable developers to be expressive with their code. To further enhance this expressivity and the flexibility of the type system, this RFC proposes the design and implementation of user-defined type functions, a feature that will allow developers to define custom type functions.

## Motivation

The primary motivation for introducing user-defined type functions is to extend the flexibility of the type system and empower developers to create more expressive and type-safe libraries. Current limitations in the type system restrict custom type manipulations, which are crucial for creating advanced patterns and abstractions. By supporting user-defined type functions, we aim to:
- Enable more precise type definitions that can capture complex relationships and constraints
- Facilitate the creation of reusable and composable libraries
- Enhance code safety and maintainability with better type-level programming

The expected outcome is a more versatile type system that can adapt to a wider range of programming patterns, ultimately improving the Luau developer experience.

## Design

> Type functions wil be defined with the following syntax:

Reviewer 3: ```suggestion
Type functions will be defined with the following syntax:
```

end
```
As a result, the current plan is to maintain a tightly scoped implementation, focusing on the simplest version of type functions to minimize the risk of developers accidentally starving themselves of features provided by the analysis, and gradually enhance capabilities through successive iterations.

We have considered implementing a user-configurable execution limit based on time or instruction count for reducing type functions where if a type function exceeds the limit, it fails to reduce. However, these methods are not consistently reliable for terminating type functions that take too long to execute. Time-based timeouts are dependent on CPU performance; programs that type check on fast CPUs may not type check on slower CPUs. Similarly, instruction-based timeouts are dependent on the compiler optimizations; programs that type check on one compiler version may not type check on versions without the same optimizations. We will be experimenting with various forms of limiting the execution of type functions and reserve the right to change how termination of type functions is managed.

<details><summary>List of illegal constructs in type functions</summary>

> * `while` / `repeat` loops

Reviewer 3: What's the reasoning? Seems very arbitrary.

Reviewer 2: @alexmccord was concerned about infinite looping potential and so forth. Personally, I'd prefer to restrict the language as little as possible, so if consensus is not to have such a limit, I'd be happy for that.

Reviewer: Yeah. I wanted to go with Starlark approach wherein we have no such case of a halting problem.
1. No infinite loops
2. `for i = 1, math.huge do` would error.
3. `for i, v in t do table.insert(t, 0) end` also errors
And some other cases like those, and I think I still want to see a path towards type functions being guaranteed to terminate in finite time.

Reviewer 5: That's why such a thing is done using "symbolic execution" everywhere, so placing such limits kind of goes against the idea of how it's better to rely on existing language execution implementation (not sure how the insert one can even be done without a different interpreter)
And what about `for i = 1, halting_problem_constant_computation` or `for i = 1,100,1e-6`, are we computing loop trip count and placing a limit?
I'm not a stranger to having multiple runtimes in a single language with limits, it's not that bad https://github.com/WheretIB/nullc/blob/master/NULLC/ExpressionEval.h#L16-L37 
On multiple representation levels even https://github.com/WheretIB/nullc/blob/master/NULLC/InstructionTreeVmEval.h#L20-L40

If we want such strict limits, including limited environment (meaning limited stuff to re-implement, which is nice), it's not a bad choice. And I heard there is a pretty in-depth constant-eval implementation on Luau Ast...

Ignoring limits is also a path. We have circle-lang having compile-time calls to git executable https://github.com/seanbaxter/circle/blob/master/examples/README.md#a-version-printer and they even have Lua interpreter built-in https://github.com/seanbaxter/circle/blob/master/examples/README.md#querying-lua
Interesting that they have all that with an interpretation over AST and not by running binary programs (modulo FFI)

Reviewer: Yeah, numeric loops would be bounded by a number of possible steps, not just `from` and `to`. In this case, you need an entirely new VM that, depending on the construct, you place some checks to ensure a limit is not being exceeded.

So basically
```luau
for i = 1, 100, 1e-6 do
end
```

becomes equivalent to

```luau
if (to - from)/step >= MY_ITERATION_LIMIT then
  error("too many iterations")
else
  for i = from, to, step do
    -- the code
  end
end
```

Reviewer: Oh, and also this:

```luau
type function oops()
  while tru| do
  end
end

function f(x: oops) end
```

If your cursor is at `|` and you type `e`, the whole analysis hangs the thread.

Reviewer 3: LSPs should implement a timeout, I know ours will, so this will not be a problem.
Hitting this timeout will also be a warning that limitless luau-analysis later on the same code might hang.

Reviewer: In that case, let's do the consistent thing and remove this limitation.

* invoking other type functions / regular functions / lambdas
    * we will not (and probably never) allow type functions to call regular functions for the sake of maintaining discrete stages between runtime and analysis
* referring to locals / globals in the outer scope
* global functions: `getfenv`, `setfenv`, `pcall`, `xpcall`, `require`
* libraries: `coroutine`, `debug`, `string.gsub`

Note: we are aware that for loops can cause infinite runtime. For the time being, we will not be handling this case. In the event that a developer accidentally creates an infinitely long type function, autocomplete will timeout in their editor environments and running luau-analyze will not complete. They will need to fix the type function and restart their environment / analysis.

</details>

> To give warnings, developers can use `print()` with custom warning messages. To fail reductions, developers can use `error()` with custom error messages. If nothing is returned by the type function, it will fail to reduce with the default message: "Failed to reduce \<Name\> type function with no return value". If the return value of the type function is not an instance of typelib, it will fail to reduce with the message: "Failed to reduce \<Name\> type function with non-typelib return value".

Reviewer 3: Can you explain the concept of a 'warning' better?
Today, there is no system for warning messages in the type inference engine (old or new).
If a new system is added, please describe the design of the warning system.

Reviewer 2: I think right now it'd probably end up working exactly the same as everything does: just producing an error message either way, but there's a lot of reasons we probably ought to introduce a technical distinction between warnings and errors in the future. That's probably just another RFC altogether though, since it's likely going to include designing things like e.g. disabling warnings both locally and globally, forcing them to be errors, etc.

Reviewer 2: Yep, I think that's a good call!

#### Union

| Instance Methods | Return Type | Description |
| ------------- | ------------- | ------------- |
| `getcomponents()` | `{typelib}` | returns an array of types that the self's union can represent. For instance, `string \| number` returns `{typelib.getstring(), typelib.getnumber()}` |

#### Intersection

| Instance Methods | Return Type | Description |
| ------------- | ------------- | ------------- |
| `getcomponents()` | `{typelib}` | returns an array of types represented by self's intersection. For instance, `string & number` returns `{typelib.getstring(), typelib.getnumber()}` |

#### Class

| Instance Methods | Return Type | Description |
| ------------- | ------------- | ------------- |
| `getprops()` | `{[typelib]: typelib}` | returns the runtime representation self's properties |
| `getparent()` | `typelib?` | returns the runtime representation of self's parent class if it exists, else nil |
| `getmetatable()` | `typelib?` | returns the runtime representation of self's metatable if it exists, else nil |
| `getindexer()` | `{key: typelib, value: typelib}?` | returns a table containing self's indexer key type and value type |

</details>

> The reason for going with userdata instead using another representation or adding a new compile-time interpreter like many other language is that userdata provides a clean abstraction from the runtime representation and restricts developers to using a set of libraries controlled by the Luau team where we can easily manage changes and listen to feedbacks. Moreover, because Lua was designed to be easily embeddable with C++, we wanted to use this as an advantage for type functions since they only require using a small runtime variant of the VM, making the implementation be simpler than having to implement a new interpreter.

Reviewer 3: ```suggestion
The reason for going with userdata instead using another representation or adding a new compile-time interpreter like many other language is that userdata provides a clean abstraction from the runtime representation and restricts developers to using a set of libraries controlled by the Luau team where we can easily manage changes and listen to feedbacks. Moreover, because Luau is designed to be easily embeddable with C++, we wanted to use this as an advantage for type functions since they only require using a small runtime variant of the VM, making the implementation be simpler than having to implement a new interpreter.
```

### Implementation

> A `typelib` library will be implemented using the Lua API to interface between C++ and Luau and support the library methods, including type serialization for arguments and deserialization for return values. To implement type functions, a new AST node called `AstStatTypeFunction` will be introduced and created when parsing a type alias followed by the keyword "function." In the constraint generator, visiting this new AST node will generate a `TypeAliasExpansionConstraint` and in the constraint solver, reducing `TypeAliasExpansionConstraint` will generate a `ReduceConstraint`. To reduce `ReduceConstraints`, user-defined type functions will be integrated as built-in type functions in Luau where when being invoked, their arguments will be serialized into an instance of `typelib`. These functions will interact with the Luau VM to execute the function body in an established environment with only the specified libraries and constructs available. The return value of the type function will be deserialized and be the value that the `ReduceConstraint` reduces to.

Reviewer 3: ```suggestion
A `typelib` library will be implemented using the Luau API to interface between C++ and Luau and support the library methods, including type serialization for arguments and deserialization for return values. To implement type functions, a new AST node called `AstStatTypeFunction` will be introduced and created when parsing a type alias followed by the keyword "function." In the constraint generator, visiting this new AST node will generate a `TypeAliasExpansionConstraint` and in the constraint solver, reducing `TypeAliasExpansionConstraint` will generate a `ReduceConstraint`. To reduce `ReduceConstraints`, user-defined type functions will be integrated as built-in type functions in Luau where when being invoked, their arguments will be serialized into an instance of `typelib`. These functions will interact with the Luau VM to execute the function body in an established environment with only the specified libraries and constructs available. The return value of the type function will be deserialized and be the value that the `ReduceConstraint` reduces to.
```

### `types` API Reference

> This section details the initial programming interface we propose for `type`s when they are reflected into type function bodies. Each section is separated by headers (e.g. Singleton) and describe the methods available to that category of type. The initial section, titled `types`, describes the elements of the interface common to every category of type. All attributes of newly-created `type`s are initialized with empty tables / arrays and `nil`. All arguments are passed by references. 

Reviewer 3: ```suggestion
This section details the initial programming interface we propose for `type`s when they are reflected into type function bodies. Each section is separated by headers (e.g. Singleton) and describe the methods available to that category of type. The initial section, titled `types`, describes the elements of the interface common to every category of type. All properties of newly-created `type`s are initialized with empty tables / arrays and `nil`. All arguments are passed by references. 
```
I think we usually refer to them as properties, considering that we have a different language feature that walks about attributes.

<details><summary>Expand for full type API reference.</summary>

> #### `types`

Reviewer 3: Maybe the next 3 tables can be formatted a bit better for clarity.

The section can start with "`types` library" where "Instance Attributes" are replaced with "Library Properties" and "Static Methods" are replaced with "Library Functions". Functions because they are called as `types.singleton(nil)`.
I don't think there's a mention to make library itself available through the `type` (like with the `string` library) and I do not suggest doing so (so no `t:unionof(types.number)`).

With this clear separation we can then make a split for `type` heading and place "Instance Methods" there.

And before moving on to `Negation` and other section we can note that `type` can have a 'tag' or a 'kind'. By the way, we have the `is` method to check the tag/kind, but there is no property to read what the kind is? This I would suggest adding.

What I mean about the structure:

#### `types` Library

| Library Properties | Type | Description |
| ------------- | ------------- | ------------- |
| `unknown` | `type` | an immutable instance of the built-in type `unknown` |
| `string` | `type` | an immutable instance of the built-in type `string` |
...

| Library Functions | Return Type | Description |
| ------------- | ------------- | ------------- |
| `singleton(arg: string \| boolean \| nil)` | `type` | returns an immutable instance of the argument as a singleton type |
...

### `type` instance

| Instance Methods | Return Type | Description |
| ------------- | ------------- | ------------- |
| `__eq(arg: type)` | `boolean` | overrides the == operator to return true if self is syntactically equal to arg |
...

Depending on the tag, `type` instance can have additional properties and methods available.

#### Negation `type`

...

Reviewer 2: Really appreciate this suggestion. Just reformatted it! 😸

| Instance Methods | Return Type | Description |
| ------------- | ------------- | ------------- |
| `__eq(arg: type)` | `boolean` | overrides the == operator to return true if self is syntactically equal to arg |
| `is(arg: string)` | `boolean` | returns true if self has the same tag as the argument. List of available tags: "nil", "unknown", "never", "any", "boolean", "number", "string", "singleton", "negation", "union", "intersection", "table", "function", "class"  |

| Static Methods | Return Type | Description |
| ------------- | ------------- | ------------- |
| `singleton(arg: string \| boolean \| nil)` | `type` | returns an immutable instance of the argument as a singleton type |
| `negationof(arg: type)` | `type` | returns an immutable instance of the parameter as a negated type; the parameter cannot be a table or function type |
| `unionof(...: type)` | `type` | returns an immutable instance of an union type of the arguments; requires at least 2 parameters to be provided |
| `intersectionof(...: type)` | `type` | returns an immutable instance of an intersection type of the arguments; requires at least 2 parameters to be provided |

> | `newtable(props: {[type]: type}?, indexer: {key: type, value: type}?, metatable: type?)` | `type` | returns a mutable instance of a table type; the keys of the table's property must be a singleton type; if provided the metatable parameter, this table becomes a metatable |

Reviewer 3: There is no 'metatable' tag in type function context. If we really want to explain the internals here (on how runtime value is converted into typechecker type), it would be best to say `MetatableType`.
But some can say that separate representation chosen for tables with metatable in the typechecker wasn't the best choice :)

> | `newtable(props: {[type]: type}?, indexer: {key: type, value: type}?, metatable: type?)` | `type` | returns a mutable instance of a table type; the keys of the table's property must be a singleton type; if provided the metatable parameter, this table becomes a metatable |

Reviewer 3: Should this be updated so that `props` match `properties()`?
If we want, we can also define these as `[type]: type | { read: type, write: type }` if we want a simplified option for common readwrite type.

| `newfunction(parameters: {pack: {type}?, variadic: type?}, returns: {pack: {type}?, variadic: type?})` | `type` | returns a mutable instance of a `function` type |
| `copy(arg: type)` | `type` | returns a deep copy of the given `type` |

#### Negation

| Instance Methods | Type | Description |
| ------------- | ------------- | ------------- |
| `inner()` | `type` | returns the `type` being negated |

#### Singleton

| Instance Methods | Return Type | Description |
| ------------- | ------------- | ------------- |

> | `value()` | `string \| boolean` | returns the string value of the  singleton |

Reviewer 3: ```suggestion
| `value()` | `string \| boolean` | returns the string value of the singleton |
```

Reviewer 3: There is a bit of imbalance between `singleton(nil)` used to create a `nil` type and `nil` type not actually being a singleton.

Reviewer 2: I think we can reasonably make `value()` work for `nil` too.

| `setreadindexer(index: type, result: type)` | `()` | sets the table's indexer with the index type as the first parameter, and the resulting read type as the second parameter |
| `setwriteindexer(index: type, result: type)` | `()` | sets the table's indexer with the index type as the first parameter, and the resulting write type as the second parameter |
| `indexer()` | `{index: type, readresult: type, writeresult: type}?` | returns the table's indexer as a table; if the indexer does not exist, returns nil |
| `readindexer()` | `{index: type, result: type}?` | returns the table's indexer as a table using the read type of the result; if the indexer does not exist, returns nil |
| `writeindexer()` | `{index: type, result: type}?` | returns the table's indexer as a table using the write type of the result; if the indexer does not exist, returns nil |
| `setmetatable(arg: type)` | `()` | sets the table's metatable to the argument |
| `metatable()` | `type?` | returns the table's metatable; if the metatable does not exist, returns nil |

#### Function

| Instance Methods | Return Type | Description |
| ------------- | ------------- | ------------- |

> | `setparameters(pack: {type}?, variadic: type?)` | `nil` | sets the function's parameter types to the arguments |

Reviewer 3: ```suggestion
| `setparameters(pack: {type}?, variadic: type?)` | () | sets the function's parameter types to the arguments |
```

> | `parameters()` | `{pack: {type}?, variadic: type?}` | returns the function's parameter types as a table |

Reviewer 3: I think type annotation gives a better description here than the description itself.

> | `setreturns(pack: {type}?, variadic: type?)` | `nil` | sets the function's return types to the arguments |

Reviewer 3: ```suggestion
| `setreturns(pack: {type}?, variadic: type?)` | () | sets the function's return types to the arguments |
```

## Motivation

The primary motivation for introducing user-defined type functions is to extend the flexibility of the type system and empower developers to create more expressive and type-safe libraries. Current limitations in the type system restrict custom type manipulations, which are crucial for creating advanced patterns and abstractions. By supporting user-defined type functions, we aim to:
- Enable more precise type definitions that can capture complex relationships and constraints
- Facilitate the creation of reusable and composable libraries
- Enhance code safety and maintainability with better type-level programming

The expected outcome is a more versatile type system that can adapt to a wider range of programming patterns, ultimately improving the Luau developer experience.

## Design

Type functions wil be defined with the following syntax:
```luau
type function f(...)
    -- implementation of the type function
end
```

For instance, the `rawget` type function can be written as:
```luau
type function rawget(tbl, prop)

>     if (not tbl:istable()) then

Reviewer 2: ```suggestion
    if not tbl:istable() then
```

        error("First parameter of rawget is not a table!") -- fails to reduce
    end

    return tbl:getprops()[prop:getvalue()]
end

type Person = {
    name: string,
    age: number
}

type ty = rawget<Person, "name"> -- ty = string
```

Type functions operate on two stages: type analysis and runtime. When calling type functions at the type level (e.g. annotating a variable as a type function), angle brackets must be used, but when calling them at the runtime level (e.g. calling other type functions within type functions), parenthesis must be used. Declarations of type functions use parenthesis because it defines the runtime operations on the runtime representation of types.

> For the first iteration, type functions will support a subset of the Luau language, mainly the constructs that do not allow type functions to run infinitely. Infinitely running type functions are problematic because they can halt the analysis from making further progress. For example, reducing this type function will recurse until a stack overflow occurs and cause the whole analysis to crash:

Reviewer 2: "Cause the whole analysis to crash" is absolutely not true. The stack overflow is in a VM instance, and you can literally read it out as the error from the VM and report to the user that the type function stack overflowed.

```luau
type function neverending(t)
    return neverending(t) -- note: parentheses are used here because the runtime value of types is being passed in, rather than the static annotation of types
end
```

> As a result, the current plan is to maintain a tightly scoped implementation, focusing on the simplest version of type functions to minimize the risk of developers crashing the analysis, and gradually enhance capabilities through successive iterations.

Reviewer 2: We're worried about developers shooting themselves (and, via dependencies, one another) in the foot with analysis time explosions, rather than concerned about crashing.

We have considered implementing a user-configurable execution limit based on time or instruction count for reducing type functions where if a type function exceeds the limit, it fails to reduce. However, these methods are not consistently reliable for terminating type functions that take too long to execute. Time-based timeouts are dependent on CPU performance; programs that type check on fast CPUs may not type check on slower CPUs. Similarly, instruction-based timeouts are dependent on the compiler optimizations; programs that type check on one compiler version may not type check on versions without the same optimizations. We will be experimenting with various forms of limiting the execution of type functions and reserve the right to change how termination of type functions is managed.

<details><summary>List of illegal constructs in type functions</summary>

* `while` / `repeat` loops
* invoking other type functions / regular functions / lambdas

>     * we will not (and probably never) allow type functions to call regular functions for the sake of sandboxing the runtime and analysis.

Reviewer 2: we will not allow this because we want discrete staging, which is a little different than sandboxing.

Reviewer: When sandboxing was brought up as an argument against this, staged execution wasn't in the picture. I think it's worth keeping both arguments against this feature because we want to both a) have staged execution and b) protect users from RCE by simply opening the code in their editor.

Reviewer 2: I think that's valid, but that's not what comes through to me from the text "sandboxing the runtime and analysis." It sounds like its trying to describe keeping them separate, as opposed to "ensuring that the type function runtime is sandboxed."

* referring to locals / globals in the outer scope
* global functions: `getfenv`, `setfenv`, `pcall`, `xpcall`, `require`

> * libraries: `coroutine`, `debug`, `string.gsub`

Reviewer: I'm thinking I want an inverse to this list. Instead of listing globals that we wish to disallow, we'll have a list of globals we wish to allow. The implementation should also reflect a list of allowed globals as well.

Don't want a type function that changes behavior based on `os.time()`.

Reviewer 3: Yeah, agree on listing what is allowed, as the current list both misses some of the items and also allows 'escapes' that were probably intended to be prevented.

> Note: we are aware that for loops can cause infinite runtime. For the time being, we will not be handling this case. In the event that a developer accidentally creates an infinitely long type function, they will need to fix the type function and restart the analysis.

Reviewer 2: The outcomes here are that they will likely guarantee autocomplete timeouts in their editor environment, and if they ran luau-analyze directly it would not complete.

<details><summary>typelib library methods (dropdown)</summary>

Methods under a different type heading (ex: `Singleton`) imply that the methods are only available for those types. At the implementation level, there is a check to make sure that the type-specific methods are being called on the correct types. For instance, `getindexer()` asserts that `istable()` is true.

#### typelib
All attributes of newly created typelib are initialized with empty tables / arrays and `typelib:getnil()`. For instance, `typelib:newtable()` initializes its properties with an empty table and index / index result type as `typelib:getnil()`.

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `getnil()` | `typelib` | returns an immutable runtime representation of the built-in type `nil` |
| `getunknown()` | `typelib` | returns an immutable runtime representation of the built-in type `unknown` |
| `getnever()` | `typelib` | returns an immutable runtime representation of the built-in type `never` |
| `getany()` | `typelib` | returns an immutable runtime representation of the built-in type `any` |

> | `getnegation(arg: typelib)` | `typelib` | returns an immutable runtime representation of the negation of the argument; the argument cannot be `istable()`, `ismetatable` or `isfunction()` |

Reviewer 2: The presence of this basically means we have to do the negation syntax RFC (since otherwise it means people will just write trivial type functions for negations). This is okay, of course, since that was already planned, but just a thing to note.

| `getboolean()` | `typelib` | returns an immutable runtime representation of the built-in type `boolean` |
| `getnumber()` | `typelib` | returns an immutable runtime representation of the built-in type `number` |
| `getstring()` | `typelib` | returns an immutable runtime representation of the built-in type `string` |
| `getstringsingleton(arg: string)` | `typelib` | returns an immutable runtime representation of a string singleton type of the argument |
| `getbooleansingleton(arg: boolean)` | `typelib` | returns an immutable runtime representation of a boolean singleton type of the argument |
| `getunion(arg: {typelib})` | `typelib` | returns an immutable runtime representation of union type of its argument |
| `getintersection(arg: {typelib})` | `typelib` | returns an immutable runtime representation of intersection type of its argument |
| `newtable()` | `typelib` | returns a mutable runtime representation of a `table` type |

> | `newmetatable()` | `typelib` | returns a mutable runtime representation of `Metatable` |

Reviewer 2: I think I'd like for metatables to be represented as a special blessed property of tables, @andyfriesen has described to me the belief that choosing otherwise (i.e. to make metatable its own sort of type) was a mistake.

| `isboolean()` | `boolean` | returns true if self is a runtime representation of the built-in type`boolean` |
| `isnumber()` | `boolean` | returns true if self is a runtime representation of the built-in type `number` |
| `isstring()` | `boolean` | returns true if self is a runtime representation of the built-in type `string` |
| `isstringsingleton()` | `boolean` | returns true if self is a runtime representation of a string singleton |
| `isbooleansingleton()` | `boolean` | returns true if self is a runtime representation of a boolean singleton |
| `isunion()` | `boolean` | returns true if self is a runtime representation of the union type |
| `isintersection()` | `boolean` | returns true if self is a runtime representation of the intersection type |
| `istable()` | `boolean` | returns true if self is a runtime representation of a `table` type |
| `ismetatable()` | `boolean` | returns true if self is a runtime representation of `Metatable` |
| `isfunction()` | `boolean` | returns true if self is a runtime representation of a `function` type |
| `isclass()` | `boolean` | returns true if self is a runtime representation of a `class` type |

> | `issubtypeof(arg: typelib)` | `boolean` | returns true if self is a subtype or equal to arg in the type hierarchy |

Reviewer 2: I'm not necessarily averse to having this in the API, but it's worth noting that this essentially requires that this library depend on `Normalization` or provides its own because Luau's notion of subtyping is fundamentally semantic. That is probably going to be a huge blocker, and this is, by far, the most complicated thing to implement in this RFC.

> | `conformsto(arg: typelib)` | `boolean` | returns true if self is equal to arg in the type hierarchy |

Reviewer 2: I don't understand what this function means at all. `conformsto`?

Reviewer: Yeah. Thought we agreed to using `__eq` and `__le` for subtyping.

Author: > I don't understand what this function means at all. conformsto?

ah, `conformsto` will be change to `equalsto`.

> Thought we agreed to using __eq and __le for subtyping.

I remember in one of the meetings, Andy preferred to have them as methods rather than overloading operators. We could do both?

Reviewer: I think I would like to pick one or the other. I suggested `__eq` and `__le` because it felt intuitive. We already say things like `T <: U` which can be expressed in Luau terms as `t <= u`.

#### Negation

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `gettype()` | `typelib` | returns the runtime representation of the type being negated |

#### String

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `getmetatable()` | `typelib` | returns the runtime representation of `Metatable` |

#### StringSingleton

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `getvalue()` | `string` | returns a string singleton |

#### BooleanSingleton

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `getvalue()` | `boolean` | returns either `true` or `false` |

#### Table

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `addprops(key: typelib, value: typelib)` | `nil` | adds a key, value pair to self's table properties; if the same key exists already, overrides the value |

> | `delprops(key: typelib)` | `nil` | removes the key from self's table properties along with the value associated with it; if the key doesn't exist, nothing happens |

Reviewer 2: these functions take singular properties, so they should be named `addprop` and `delprop`

Reviewer: We could use `__newindex` for this.

```luau
t["myprop"] = mytype -- adds the prop
t["myprop"] = nil -- deletes the prop
```

Reviewer 2: Yeah, that would be considerably nicer.

Reviewer 2: That works, in my mind, if we actually make `{ [ "foo" | "bar" | "baz"]: string }` equivalent to `{ foo: string, bar: string, baz: string }`, but I don't think it's so bad if indexers are an explicit thing.

| `getprops()` | `{typelib: typelib}` | returns a table of self's table properties (e.g. `{["age"] = 20}` will return `{typelib:getstringsingleton("age") = typelib:getnumber()}`) |
| `settableindexer(indexty: typelib, indexresultty: typelib)` | `nil` | sets self's index type to the first argument and index result type to the second |
| `getindextype()` | `typelib` | returns self's index type |
| `getindexresulttype()` | `typelib` | returns self's index result type |
| `setsealed(arg: boolean)` | `nil` | sets the table's sealed status. If true, the table becomes a sealed table. Default initialization of tables is unsealed tables (aka setsealed(false)) |

> | `issealed()` | `boolean` | returns the seal status of self |

Reviewer 2: I don't think either of these should be exposed to users at all.

Reviewer: Yeah. Absolutely an implementation detail.

#### Metatable

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `settable(arg: typelib)` | `nil` | sets self's table to the argument; the argument needs to be `istable()` |
| `gettable()` | `typelib` | returns self's runtime representation of the type `table` |
| `setmetatable(arg: typelib)` | `nil` | sets self's metatable to the argument; the argument needs to be `ismetatable()` |

> | `getmetatable()` | `typelib` | returns self's runtime representation of `Metatable` |

Reviewer 2: Just reiterating here that I think I'd rather metatables just be a property on tables, so they'd have `setmetatable` and `getmetatable` there, and not have this layer of indirection (I know that luau's analysis _does_ implement it that way).

## Drawback

Type functions are handled at the analysis time, while `typelib` is an implementation in the runtime. As a result, the proposed design causes the analysis time to be dependent on Luau's runtime to reduce user-defined type functions. This is generally discouraged as it is best to isolate the compile time, analysis time, and runtime from each other for the purpose of maintaining a clean separation of concerns, which helps minimize side effects and dependencies across different phases of the program execution and improves the modularity of the compiler. Overlaps between the analysis time and runtime can lead to code that is more complex and harder to manage, potentially increasing the risk of bugs and making the outcomes less predictable.

> The build / analysis times will also be negatively impacted by this feature as reducing type functions takes variable amount of time based on the program. The larger the type function, the longer it will take to reduce it in the constraint solver.

Reviewer 2: This is kind of complicated, it's not really a function of the size, but rather just that programmers can write non-performant code and that will impact their (and any of their dependent code's) analysis time. There is a size effect in that we have to parse things, and that'll be parsed in the bytecode compiler as well, but that's pretty normal.

- Enable more precise type definitions that can capture complex relations on types
- Facilitate the creation of type-safe, reusable and composable libraries
- Enhance Luau's support for type-level programming

The expected outcome is a more versatile type system that can adapt to a wider range of programming patterns, improving the Luau developer experience.

## Design

Type functions will be defined with the following syntax:
```luau
type function f(...)
    -- implementation of the type function
end

> ```

Reviewer 3: Section on the syntax could use more details to answer questions about visibility, scoping and shadowing.

I assume it will be similar to types?
As in, all type functions defined in the current scope including parent scopes are visible regardless is they are defined above or below each other.
With types, shadowing is not allowed (so if we introduce a new built-in type function, it will cause type errors in code that happens to use the same name). Maybe this can be relaxed to only disallow shadowing in the single scope?
Local type definitions are allowed and so are the type functions? That will imply the VM type function set has to be more dynamic instead of just collecting all type functions from outer scope once.

For parsing purposes, the type functions are parsed with an empty variable environment, similar to how they are compiled to global accesses.

Reviewer 2: Yeah, I can add this information to the RFC.

I think what I originally had in mind is: Every type alias and type function would be defined in an order-independent way (i.e. we parse them all just as we do ordinary functions, and then we build up a mapping from symbols to their definitions to be used by the type system in invoking them). This would suggest a scoping rule that is essentially "one global scope for types, plus any nested runtime scopes within type function bodies." I think this should match how the existing "type functions" work, i.e. when you write `type

Reviewer 3: Oh, to be more precise about current type shadowing rules, you are not allowed to shadow built-in types, but you can shadow a type from the global/env scope (at Roblox, `type Instance = number` will be ok, this way we freely add new environment types).
Since there are only built-in type functions and not additional ones injected from definition files (yet?), shadowing of built-ins will always be an error.

Reviewer 2: Really, I'd just like it to be consistent with everything type aliases already do, so I suppose from that perspective we should probably just say we'll support scoping and so forth in the same way. Consistent scoping rules are good, and nobody has time to keep track of weird idiosyncratic additional ones.

## Motivation

The primary motivation for introducing user-defined type functions is to extend the flexibility of the type system and empower developers to create more expressive and type-safe libraries. Current limitations in the type system restrict custom type manipulations, which are crucial for creating advanced patterns and abstractions. By supporting user-defined type functions, we aim to:
- Enable more precise type definitions that can capture complex relationships and constraints
- Facilitate the creation of reusable and composable libraries
- Enhance code safety and maintainability with better type-level programming

The expected outcome is a more versatile type system that can adapt to a wider range of programming patterns, ultimately improving the Luau developer experience.

## Design

Type functions wil be defined with the following syntax:
```luau
type function f(...)
    -- implementation of the type function
end
```

For instance, the `rawget` type function can be written as:
```luau

> type function rawget(tbl, prop)

Reviewer: One thing I'm concerned with is that users might write a type function implementations without handling union types. This example is a perfect one: `rawget` is distributive over `tbl` and `prop`, but the implementation doesn't allow for it.

Do we need a `@distributive` attribute for type functions to opt in? Or the inverse `@nondistributive`.

Mostly thinking if there was a way to lift this logic to analysis or compiler that would handle this for the users.

Reviewer: Or we could add higher kinded types, `distribute<rawget, { p: "a", q: "x" } | { p: "b", q: "y" }, "p" | "q">`

Author: i feel like this should be left as the developers job. we give all the tools they need, but not necessarily help them build whatever they want to build. maybe catching the union case can be left up for the autocomplete?

Reviewer 2: I think it should be comfortable to write the case dealing with unions (e.g. we support iterating over unions), but I don't think we need to be worried about fully automating this stuff away. We ought to wait to see how people use type functions, and what patterns are common and uncommon before we start adding abstractions to optimize writing them.

    * e.g. `t1 == t2` is equivalent to `t1:equalsto(t2)`

#### Negation

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `gettype()` | `typelib` | returns the runtime representation of the self's type being negated |

#### String

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `getmetatable()` | `typelib` | returns the runtime representation of self's metatable |

#### StringSingleton

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `getvalue()` | `string` | returns self's value of a string singleton |

#### BooleanSingleton

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `getvalue()` | `boolean` | returns self's boolean singleton value of either `true` or `false` |

#### Table

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `addprop(key: typelib, value: typelib)` | `nil` | adds a key, value pair to self's table properties; if the same key exists already, overrides the value |

> | `delprop(key: typelib)` | `nil` | removes the key from self's table properties along with the value associated with it (equivalent of `addprop(key, nil)`); if the key doesn't exist, nothing happens |

Reviewer: I would like to see this removed in favor of `__newindex`.

Author: for readers: the reason we are not doing this override is that `__newindex` and `__index` conflicts with methods if a developer so happens to have a property name that matches one of the method

> | `getprops()` | `{[typelib]: typelib}` | returns a table of self's table properties (e.g. `{["age"] = 20}` will return `{typelib.getstringsingleton("age") = typelib.getnumber()}`) |

Reviewer: Can this be removed in favor of `__index`? For iterating over tables, we can provide `__iter` as well.

```luau
local foo = t["foo"]
local bar = t.bar

for k, v in t do -- k : ty, v : ty
  -- ...
end
```

Reviewer: hm. I've just realized that this conflicts with `__iter` on `ty` itself (i.e. all types are iterable), which is something I wanted to see.

Then for the `rawget` example, it's easy to distribute type arguments.

```luau
type function rawget(t, p)
  local res = {}
  for _, t in t do
    for _, p in p do
      assert(t:istable())
      table.insert(res, t:getprops()[p])
    end
  end
  return ty.union(res)
end

-- "a" | "b" | "x" | "y"
type T = rawget<{ p: "a", q: "x" } | { p: "b", q: "y" }, "p" | "q">
```

Reviewer: It also invites a lot of nested loops per type parameter.

| `setindexer(key: typelib, value: typelib)` | `nil` | sets self's indexer key type to the first argument and indexer value type to the second |
| `getindexerkeytype()` | `typelib` | returns self's indexer key type |

> | `getindexervaluetype()` | `typelib` | returns self's indexer value type |

Reviewer: I think we can unify these two functions into one, and return some value of type `{ key: ty, value: ty }?`.

| `getparameters()` | `{typelib} \| typelib` | returns the runtime representation of self's parameter type |
| `setreturns(arg: {typelib} \| typelib)` | `nil` | sets self's return types to the argument, where an array implies a TypePack and the latter implies a Variadic |
| `getreturns()` | `{typelib} \| typelib` | returns the runtime representation of self's return type |

#### Union

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `getcomponents()` | `{typelib}` | returns an array of types that the self's union can represent. For instance, `string \| number` returns `{typelib.getstring(), typelib.getnumber()}` |

#### Intersection

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `getcomponents()` | `{typelib}` | returns an array of types represented by self's intersection. For instance, `string & number` returns `{typelib.getstring(), typelib.getnumber()}` |

#### Class

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |

> | `getprops()` | `{[typelib]: typelib}` | returns the runtime representation self's properties |

Reviewer: Ditto https://github.com/luau-lang/rfcs/pull/45/files#r1663222683.

| `getparent()` | `typelib \| nil` | returns the runtime representation of self's parent class if it exists, else nil |
| `getmetatable()` | `typelib \| nil` | returns the runtime representation of self's metatable if it exists, else nil |
| `getindexerkeytype()` | `typelib` | returns self's indexer key type |

> | `getindexervaluetype()` | `typelib` | returns self's indexer value type |

Reviewer: Ditto.

<details><summary>List of illegal constructs in type functions</summary>

* `while` / `repeat` loops
* invoking other type functions / regular functions / lambdas
    * we will not (and probably never) allow type functions to call regular functions for the sake of sandboxing the runtime and analysis.
* referring to locals / globals in the outer scope
* global functions: `getfenv`, `setfenv`, `pcall`, `xpcall`, `require`
* libraries: `coroutine`, `debug`, `string.gsub`

Note: we are aware that for loops can cause infinite runtime. For the time being, we will not be handling this case. In the event that a developer accidentally creates an infinitely long type function, they will need to fix the type function and restart the analysis.

</details>

To give warnings, developers can use `print()` with custom warning messages. To fail reductions, developers can use `error()` with custom error messages. If nothing is returned by the type function, it will fail to reduce with the default message: "Failed to reduce \<Name\> type function with no return values".

> To allow Luau developers to modify the runtime values of types in type functions, this RFC proposes introducing a new userdata called `typelib`. A `typelib` object is a runtime representation of all types within the program and provides a basic set of library methods that can be used to modify types. As such, under the hood, the `typelib` library will closely mimic the implementation of static types in Luau Analysis. Most importantly, they are *only accessible within type functions* and are *not a runtime type for other use cases than type functions*. 

Reviewer: Hate to bikeshed the name again, but could we call them simply `ty`? The clash with `type` global is unfortunate, but I think `typelib` is weirder than `ty` when no libraries we offer have `lib` in the name.

So essentially `s/typelib/ty/`.

```suggestion
To allow Luau developers to modify the runtime values of types in type functions, this RFC proposes introducing a new userdata called `ty`. A `ty` object is a runtime representation of all types within the program and provides a basic set of library methods that can be used to modify types. As such, under the hood, the `ty` library will closely mimic the implementation of static types in Luau Analysis. Most importantly, they are *only accessible within type functions* and are *not a runtime type for other use cases than type functions*. 
```

Reviewer 2: When I proposed `typelib` it was also only intended to be the name of the _library_, not the name of the userdata (which I wanted to just be called `type`) anyway. Really, I kind of just want to use `type` and overload `__call` to be the builtin `type`, but barring that, I'm happy with `ty`. Other bikeshed possibilities: `rttype`, `rtty`, `typ`.

Reviewer: Overloading `__call` on a library is not farfetched. `vector` RFC wants to do this, and no one seems to bat an eye.

If you say `type(type)` you'll get `"function"` in normal context and `"table"` in type function context, but no one should be concerned with that.

<details><summary>typelib library methods (dropdown)</summary>

Methods under a different type heading (ex: `Singleton`) imply that the methods are only available for those types. At the implementation level, there is a check to make sure that the type-specific methods are being called on the correct types. For instance, `getindexer()` asserts that `istable()` is true.

#### typelib
All attributes of newly created typelib are initialized with empty tables / arrays and `typelib.getnil()`. For instance, `typelib.newtable()` initializes its properties with an empty table and index / index result type as `typelib.getnil()`.

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `getnil()` | `typelib` | returns an immutable runtime representation of the built-in type `nil` |
| `getunknown()` | `typelib` | returns an immutable runtime representation of the built-in type `unknown` |
| `getnever()` | `typelib` | returns an immutable runtime representation of the built-in type `never` |

> | `getany()` | `typelib` | returns an immutable runtime representation of the built-in type `any` |

Reviewer: Can we offer them as a _field_ of the library, not a function? e.g. `math.huge` is a field of `math`.

Obviously you'd have to freeze the `ty` library, but that's expected. `math.huge = 2` is already an error.

Reviewer 2: Yeah, if it's just a single value, it should be a field, not a function.

Author: do we need `typelib.nil` / can we use nil instead? for example,
`tbl.setprop(myKey, typelib.nil)` vs `tbl.setprop(myKey, nil)`

Reviewer 2: Those mean entirely different things to me. `nil` is already a runtime value, but `typelib.niltype` is the runtime value of the _type_ of `nil`.

Reviewer: > Those mean entirely different things to me. `nil` is already a runtime value, but `typelib.niltype` is the runtime value of the _type_ of `nil`.

Was this replied to the wrong thing?

Reviewer 2: I think @joonyoo181 deleted the comment? but if not, yeah, I must've clicked the wrong one.

| `gettype()` | `typelib` | returns the runtime representation of the self's type being negated |

#### String

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `getmetatable()` | `typelib` | returns the runtime representation of self's `Metatable` |

#### StringSingleton

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `getvalue()` | `string` | returns self's value of a string singleton |

#### BooleanSingleton

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `getvalue()` | `boolean` | returns self's boolean singleton value of either `true` or `false` |

#### Table

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `addprops(key: typelib, value: typelib)` | `nil` | adds a key, value pair to self's table properties; if the same key exists already, overrides the value |
| `delprops(key: typelib)` | `nil` | removes the key from self's table properties along with the value associated with it; if the key doesn't exist, nothing happens |

> | `getprops()` | `{typelib: typelib}` | returns a table of self's table properties (e.g. `{["age"] = 20}` will return `{typelib.getstringsingleton("age") = typelib.getnumber()}`) |

Reviewer: Presumably you mean
```suggestion
| `getprops()` | `{[typelib]: typelib}` | returns a table of self's table properties (e.g. `{["age"] = 20}` will return `{[typelib.getstringsingleton("age")] = typelib.getnumber()}`) |
```

> | `settableindexer(indexty: typelib, indexresultty: typelib)` | `nil` | sets self's index type to the first argument and index result type to the second |

Reviewer: It is implied to be a `table` already, so this duplication is redundant.
```suggestion
| `setindexer(key: typelib, value: typelib)` | `nil` | sets self's index type to the first argument and index result type to the second |
```

| `getindextype()` | `typelib` | returns self's index type |

> | `getindexresulttype()` | `typelib` | returns self's index result type |

Reviewer: I've regretted the verbose names I've given to the `TableIndexer` fields. I wanted to get around to naming them `keyType` and `valueType`.

I want to avoid making this mistake in userspace.

| `issealed()` | `boolean` | returns the seal status of self |

#### Metatable

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `settable(arg: typelib)` | `nil` | sets self's table to the argument; the argument needs to be `istable()` |
| `gettable()` | `typelib` | returns self's runtime representation of the type `table` |
| `setmetatable(arg: typelib)` | `nil` | sets self's metatable to the argument; the argument needs to be `ismetatable()` |
| `getmetatable()` | `typelib` | returns self's runtime representation of `Metatable` |

#### Function

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `setargtypes(arg: {typelib} \| typelib)` | `nil` | sets self's argument types to the argument, where an array implies a TypePack and the latter implies a Variadic |
| `getargtypes()` | `{typelib} \| typelib` | returns the runtime representation of self's argument type |
| `setreturntypes(arg: {typelib} \| typelib)` | `nil` | sets self's return types to the argument, where an array implies a TypePack and the latter implies a Variadic |

> | `getreturntypes()` | `{typelib} \| typelib` | returns the runtime representation of self's return type |

Reviewer: There's a footgun here: rank-N. If you create a type parameter and you wish to create a function type with that type parameter, then the type parameter will not be instantiated when the function is called.

You would need `addtypeparam`, but now a new footgun: what if you forgot to call it.

Reviewer: Another footgun: calling `addtypeparam` on two functions to form `<a>(a) -> <a>(a) -> a` where `<a>` in both function types here are the _same_ type parameter.

> | `setargtypes(arg: {typelib} \| typelib)` | `nil` | sets self's argument types to the argument, where an array implies a TypePack and the latter implies a Variadic |
> | `getargtypes()` | `{typelib} \| typelib` | returns the runtime representation of self's argument type |
> | `setreturntypes(arg: {typelib} \| typelib)` | `nil` | sets self's return types to the argument, where an array implies a TypePack and the latter implies a Variadic |
> | `getreturntypes()` | `{typelib} \| typelib` | returns the runtime representation of self's return type |

Reviewer: Nitpick.
```suggestion
| `setparameters(arg: {typelib} \| typelib)` | `nil` | sets self's argument types to the argument, where an array implies a TypePack and the latter implies a Variadic |
| `getparameters()` | `{typelib} \| typelib` | returns the runtime representation of self's argument type |
| `setreturns(arg: {typelib} \| typelib)` | `nil` | sets self's return types to the argument, where an array implies a TypePack and the latter implies a Variadic |
| `getreturns()` | `{typelib} \| typelib` | returns the runtime representation of self's return type |
```

#### Union

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |
| `getoptions()` | `table` | returns an array of types that the self's union can represent. For instance, `string \| number` returns `{typelib.getstring(), typelib.getnumber()}` |

#### Intersection

| Function Declaration | Return Type | Description |
| ------------- | ------------- | ------------- |

> | `getparts()` | `table` | returns an array of types represented by self's intersection. For instance, `string & number` returns `{typelib.getstring(), typelib.getnumber()}` |

Reviewer: Can we rename them to `components` or `constituents` or something? The names `parts` and `options` are an implementation detail in my opinion.

Reviewer: Also, we should offer `__iter` metamethod for them. If you have a union, you absolutely want to iterate over it. No other way around it.

type ty = rawget<Person, "name"> -- ty = string
```

> Type functions operate on two stages: type analysis and runtime. They are _called_ at the type level of Luau, but _runs_ at runtime to operate on the values of types (such as the arguments of the function). For this reason, type functions are declared using parenthesis and invoked using angle brackets. Moreover, every type functions will be part of and ran within the same scope to allow type functions to call other type functions. We will not (and probably never) allow type functions to call regular functions for the sake of sandboxing the runtime and analysis time. There will also be a user-configurable timeout for type functions based on the number of instructions. For Roblox, this could be done by editing the config file for RCC and Studio and for non-Roblox users, this could be done by setting a flag when running the compiler.

Reviewer 3: How will this instruction limit be implemented?

Reviewer 3: I there is an option to force single-stepping the interpreter using an option commonly used for debuggers.

It should be mentioned why the time limit is not used (it's used for time limits in other runtime scenarios).

While it could be said that time limit is too dependent on the CPU, instruction limit is also not fixed as updates to the compiler can change the bytecode and cause previously typechecked code to fail (should be noted).

Reviewer: ah, the halting problem, our old friend.

Reviewer: I think instruction count is ok. The issue with doing wall clock measurement is that it's dependent on CPU. At least with instruction count it's guaranteed to be consistent across CPUs, even though it's not guaranteed to be consistent across compiler versions. This is nevertheless an improvement over wall clock timeout.

Reviewer 2: Yeah, having used some type systems that rely on SMT solvers with a wall clock timeout, I really don't want to deal with inconsistencies across CPUs, and an amorphous target for developers of "optimize this to be fast enough always." At least with instruction counts, if it breaks, there is a very clear line for what you need to do to fix it again.

Author: update: an instruction limit will not be implemented. type functions will not allow any constructs that can cause infinite runtime

Reviewer: Unfortunately we're gonna have to go back to doing instruction limit. It's not as simple as "make sure certain AST nodes doesn't occur in this context."

```luau
for i = 1, math.huge do
end
```
```lua
local t = {1}
for _ in t do
  table.insert(t, 1)
end
```

Reviewer: I think I want to see wordings that essentially says "_potentially_ infinite loops in type function context are not supported." The idea I want to get across is that we reserve the right to modify the runtime semantics at any point that makes type functions throw a runtime error if infinite loops are found, which would cause these type functions to not reduce.

Meaning that at some point, we may add code that detects whether `table.insert` is being used on a table with an active iterator, and have them throw a runtime error.

> Type functions operate on two stages: type analysis and runtime. They are _called_ at the type level of Luau, but _runs_ at runtime to operate on the values of types (such as the arguments of the function). For this reason, type functions are declared using parenthesis and invoked using angle brackets. Moreover, every type functions will be part of and ran within the same scope to allow type functions to call other type functions. We will not (and probably never) allow type functions to call regular functions for the sake of sandboxing the runtime and analysis time. There will also be a user-configurable timeout for type functions based on the number of instructions. For Roblox, this could be done by editing the config file for RCC and Studio and for non-Roblox users, this could be done by setting a flag when running the compiler.

Reviewer 3: I don't think this RFC should mention what and how Roblox is going to do anything.

> Type functions operate on two stages: type analysis and runtime. They are _called_ at the type level of Luau, but _runs_ at runtime to operate on the values of types (such as the arguments of the function). For this reason, type functions are declared using parenthesis and invoked using angle brackets. Moreover, every type functions will be part of and ran within the same scope to allow type functions to call other type functions. We will not (and probably never) allow type functions to call regular functions for the sake of sandboxing the runtime and analysis time. There will also be a user-configurable timeout for type functions based on the number of instructions. For Roblox, this could be done by editing the config file for RCC and Studio and for non-Roblox users, this could be done by setting a flag when running the compiler.

Reviewer 3: RFC should not use 'could'. It's a specification.

type ty = rawget<Person, "name"> -- ty = string
```

Type functions operate on two stages: type analysis and runtime. When calling type functions at the type level (e.g. annotating a variable as a type function), angle brackets must be used, but when calling them at the runtime level (e.g. calling other type functions within type functions), parenthesis must be used. Declarations of type functions use parentheses because it defines the runtime operations on the runtime representation of types.

For the first iteration, the body of a type function will be sandboxed, and its scope will be limited, meaning it will be unable to refer statements defined in the outer scope, including other type functions. Additionally, type functions will be limited on what globals/libraries they could call. The list of available globals/libraries in type functions is:

- global functions: `assert`, `error`, `next`, `print`, `rawequal`, `select`, `tonumber`, `tostring`, `type`, `typeof`, `ipairs`, `pairs`, `unpack`
- math library
- table library
- bit32 library
- buffer library

> There is also a problem of infinitely running type functions. For example, reducing this type function will halt analysis until the VM stack overflows:

Reviewer: There's a couple types of termination that's important to be pedantic here:
1) recursion
2) exception
3) infinite loop

For both 1 and 2, both of them are guaranteed to throw a runtime error and the VM halts. For 3, the VM will never halt. This snippet uses recursion, so "unbounded recursion" is better than "infinite" here, but since the following context talks about enforcing time limits, you really want this snippet to have some infinite loop rather than recursion (especially since Luau doesn't presently implement TCO).

| `issubtypeof(arg: typelib)` | `boolean` | returns true if self is syntactically a subtype or equal to arg in the type hierarchy |
| `equalsto(arg: typelib)` | `boolean` | returns true if self is syntactically equal to arg in the type hierarchy |

* `issubtypeof(arg: typelib)` and `equalsto(arg: typelib)` will also overload `__le` and `__eq` respectively
    * e.g. `t1 <= t2` is equivalent to `t1:issubtypeof(t2)`

>     * e.g. `t1 == t2` is equivalent to `t1:equalsto(t2)`

Reviewer: I know we said to delete the `<=`, but I thought we compromised by keeping `==` and changing `<=` to `issubtypeof`.

| Static Methods | Return Type | Description |
| ------------- | ------------- | ------------- |

> | `singleton(arg: string \| boolean \| nil)` | `type` | returns an immutable instance of the argument as a singleton type |

Reviewer: I like `scalar` the best, but fine with this name if we want to keep it.

Reviewer 2: `singleton` is what `is` checks against as a string.

| Instance Methods | Type | Description |
| ------------- | ------------- | ------------- |
| `inner()` | `type` | returns the `type` being negated |

#### Singleton

| Instance Methods | Return Type | Description |
| ------------- | ------------- | ------------- |
| `value()` | `string \| boolean` | returns the string value of the  singleton |

#### Table

| Instance Methods | Return Type | Description |
| ------------- | ------------- | ------------- |
| `setproperty(key: type, value: type?)` | `nil` | adds / overrides (if same key exists) a key, value pair to the table's properties; if value is nil, removes the key, value pair; if the key does not exist and the value is nil, nothing happens |

> | `getproperty(key: type)` | `type?` | returns the value associated with the key from the table's properties; if the key does not exists, returns nil |

Reviewer: Most of the API clipped `get` prefix which is a big +1 from me, but this one is an odd duck.

Reviewer 2: I had a comment I was writing earlier (that I'll probably post tomorrow when I'm back on my work PC) about the change, but the lua libraries themselves are inconsistent on this kind of thing, and... frankly, I tried to mirror what I imagine the rationale to be here too. I don't mind calling it `property` too much, but it does feel a little bit off to me because of it having a parameter, I think.

> | `properties()` | `{[type]: type}` | returns a table of the table's properties |

Reviewer: Tables without `__newindex` metamethod can still have new properties added. The question is basically: if I call `:properties()` do I get a reference to the internal table, or do I get a copy of the internal table? If it's a reference which I assume it is, then it allows you to bypass any validation that might occur as part of `:setproperty()`.

> | `properties()` | `{[type]: type}` | returns a table of the table's properties |

Author: oh, i see.. in that case developers can modify class properties :/. and yes, everything is pass by references unless someone calls the `copy()` method. im not sure how we could mitigate that vulnerability

> | `properties()` | `{[type]: type}` | returns a table of the table's properties |

Reviewer 2: I would expect the table to be frozen, I think, or at least divorced from the actual underlying representation of the table type. Though, if we could make index, iter, and so forth all work right on table types, I think I'd like that quite a bit too, but I remember you expressing some concern about that somewhere in this sea of hundreds of comments.

end
```

> In some sense, this design could be considered "cleaner" than introducing an entirely new userdata, but it requires developers to have a deeper understanding of the type runtime representation. For example, under the proposed design, a new table type can be created using `lType.new("table")`, while under the alternative design, tables must be declared with attributes: `{type = "table", props = {}, indexer = {}}`. This adds complexity to the developer experience and increases the chance of making syntax errors in their program.

Reviewer: > it requires developers to have a deeper understanding of the type runtime representation

They don't need to. If you're authoring a type function, it's very likely that you have autocomplete setup, right? We'd expose the details in autocomplete.

Author: I still think it is worth mentioning this just to be on the "safer" side. We can't really assume anything about what environments the developers may be using right?
